Code generation documentation

gonzodepedro opened the pull request.

mergify[bot] committed:
Replace dead link in Creating-A-Workspace.rst (#4148) (#4150)

Signed-off-by: Romain Lebbadi-Breteau <romain@lebbadi.fr>
(cherry picked from commit 8fd58229c345189a46caa886d3ed830e542dc019)

Co-authored-by: Romain Lebbadi-Breteau <romain@lebbadi.fr>

tfoote committed:
Add mermain support and update pinned version

To get mermaide support working I had to do upgrades. I've updated our pins and documented how they're used as well as how to update them too.

This will allow us to do integrated mermaid sequence diagrams as well as flow diagrams instead of committing images of them.

tfoote committed:
Copy over Metrics from wiki to ros2_documentation (#4147)

These reports are important to continue to provide visiblity for ROS 2 as well as maintain the history.

Migrating from: https://wiki.ros.org/Metrics

tfoote committed:
Match folders in ignore too

Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>
Signed-off-by: Tully Foote <tully.foote@gmail.com>

tfoote committed:
Unoverlap lines

Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>
Signed-off-by: Tully Foote <tully.foote@gmail.com>

tfoote commented:
I got mermaid working in: https://github.com/ros2/ros2_documentation/pull/4151 that you can build on top of.

MichaelRazum committed:
Update Setting-Up-Simulation-Webots-Advanced.rst (#4146)

The import statement `from launch_ros.actions import Node` was missing in line 3

Signed-off-by: Michael Razum <michael.razoumovitch@gmail.com>

gonzodepedro committed:
Added documentation for code generation

Signed-off-by: Gonzalo de Pedro <gonzalodepedro@ekumenlabs.com>

tfoote commented on source/Concepts/Advanced/About-Internal-Interfaces.rst:
```diff
@@ -191,3 +191,58 @@ These are mainly used for error handling, commandline argument parsing, and logg
+of CMake and Ament. To simplify the explanation of this mechanism let's separate ROS2 projects that generate or need code to be generated in two types.
+
+-  ``Generator`` projects, which are the ones that have the templates and logic to generate code.
+-  ``Client`` projects, which are the ones that need code to be generated, they have the parameters to generate code.
```
Client as a keyword collides with the "ros client library" verbiage. I might suggest an alternative to `client` to differentiate from `rcl`. I would suggest potentially `interface` projects.

gonzodepedro replied:
Change "client" for "interfaces"

tfoote commented on source/Concepts/Advanced/About-Internal-Interfaces.rst:
```diff
@@ -191,3 +191,58 @@ These are mainly used for error handling, commandline argument parsing, and logg
+Automatic code generation using template files
+==============================================
+
+As mentioned above, in order to deal with different DDS implementations, code has to be generated. Code generation is handled by a clever usage
```
```suggestion
As mentioned above, in order to deal with different DDS implementations, code has to be generated. Code generation is handled by a combination
```

tfoote commented on source/Concepts/Advanced/About-Internal-Interfaces.rst:
```diff
@@ -191,3 +191,58 @@ These are mainly used for error handling, commandline argument parsing, and logg
+    Note right of ament_package: Execute a cmake.in template with variables and a cmake code hook.
+
+
+When the *client* package, during its build process, calls the 'rosidl_generate_interfaces' macro, the extensions that were registered for each of the
```
Please wrap on sentence boundaries. That's our standard for documentation. It makes diffs easier to read.

tfoote commented on source/Concepts/Advanced/About-Internal-Interfaces.rst:
```diff
@@ -191,3 +191,58 @@ These are mainly used for error handling, commandline argument parsing, and logg
+    participant generate_arguments_file
+    participant rosidl_generator
+    project_with_rosidl_files->>rosidl_generate_interfaces: path_to_idl_files
+    loop ForEach Hook
```
It would be good to include information about how many times this will loop.

tfoote committed:
remove now redundant pycache gitignore

Signed-off-by: Tully Foote <tullyfoote@intrinsic.ai>

tfoote committed:
Bring in support for mermaid and move pinning forward (#4151)

* Add mermaid support and update pinned version

To get mermaid support working I had to do upgrades. I've updated our pins and documented how they're used as well as how to update them too.

This will allow us to do integrated mermaid sequence diagrams as well as flow diagrams instead of committing images of them.

* clean up docker installation to focus on pip version pinning

* Match folders in ignore too

Signed-off-by: Tully Foote <tullyfoote@intrinsic.ai>
Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

gonzodepedro committed:
Changed interface name

Signed-off-by: Gonzalo de Pedro <gonzalodepedro@ekumenlabs.com>

tfoote commented:
There's some more advanced features here: https://mermaid.js.org/syntax/sequenceDiagram.html
